### PR TITLE
Tidy & reword Secret concept

### PR DESCRIPTION
**Pull request by sftim**
- Tidy to match website style guide
- Expand the overview
- Use callout blocks for notes, warnings and cautions where it seems appropriate
- General rewording
- Page restructuring
   - Move all the information security details into one place
   - Consolidate details and advice about Kubernetes behavior if a Secret is missing, or if it exists but lacks a mandatory key

[_Preview_](https://deploy-preview-30984--kubernetes-io-master-staging.netlify.app/docs/concepts/configuration/secret/)

Note to reviewers:
- this PR does _not_ excise or migrate https://kubernetes.io/docs/concepts/configuration/secret/#use-cases. I think it would be good for a future PR / separate piece of work to look at moving that part of the page into the Tasks part of the documentation.

Based on #24169

/language en
/sig storage
/sig security

### CODE REVIEW

File: content/en/docs/concepts/configuration/secret.md
Comment (sftim) [inline content/en/docs/concepts/configuration/secret.md:95]: This is retained; I move it to later in the page.

File: content/en/docs/concepts/configuration/secret.md
Comment (chrisnegus) [inline content/en/docs/concepts/configuration/secret.md:529]: The link to "Add ImagePullSecrets to a service account" is repeated on about 20 lines up from here. Just want to make sure you meant to have it in both places.

File: content/en/docs/concepts/configuration/secret.md
Comment (chrisnegus) [inline content/en/docs/concepts/configuration/secret.md:539]: Consider adding text between the level 2 heading that the first level 3 heading to introduce the use cases below.

File: content/en/docs/concepts/configuration/secret.md
Comment (chrisnegus) [inline content/en/docs/concepts/configuration/secret.md:970]:
```suggestion
server checks whether the expected key exists in the `data` field, and
```

File: content/en/docs/concepts/configuration/secret.md
Comment (chrisnegus) [inline content/en/docs/concepts/configuration/secret.md:1014]:
```suggestion
- `username`: the user name for authentication
```

File: content/en/docs/concepts/configuration/secret.md
Comment (chrisnegus) [inline content/en/docs/concepts/configuration/secret.md:1015]:
```suggestion
- `password`: the password or token for authentication
```

File: content/en/docs/concepts/configuration/secret.md
Comment (chrisnegus) [inline content/en/docs/concepts/configuration/secret.md:1035]:
```suggestion
You can create an `Opaque` type for credentials used for basic authentication.
```

File: content/en/docs/concepts/configuration/secret.md
Comment (chrisnegus) [inline content/en/docs/concepts/configuration/secret.md:1036]: I don't understand what a "define Secret type" refers to.

File: content/en/docs/concepts/configuration/secret.md
Comment (chrisnegus) [inline content/en/docs/concepts/configuration/secret.md:1063]: Should this say "Opaque type" or "Opaque Secret"?

File: content/en/docs/concepts/configuration/secret.md
Comment (chrisnegus) [inline content/en/docs/concepts/configuration/secret.md:1138]:
```suggestion
well-known ConfigMaps.
```

File: content/en/docs/concepts/configuration/secret.md
Comment (chrisnegus) [inline content/en/docs/concepts/configuration/secret.md:1211]:
```suggestion
  by significantly reducing load on kube-apiserver. The kubelet does not need to maintain
```
Extraneous letter C?

File: content/en/docs/concepts/configuration/secret.md
Comment (chrisnegus) [inline content/en/docs/concepts/configuration/secret.md:1243]: I don't understand "can reason about the power of the Secrets..." Maybe it's fine. I just don't understand.

**Review by chrisnegus — COMMENTED** [review]
Very nice work @sftim ! I had very few suggestions for such a long piece.

File: content/en/docs/concepts/configuration/secret.md
Comment (sftim) [reply content/en/docs/concepts/configuration/secret.md:196]:
```suggestion
### Using Secrets as files from a Pod {#using-secrets-as-files-from-a-pod}
```

File: content/en/docs/concepts/configuration/secret.md
Comment (sftim) [reply content/en/docs/concepts/configuration/secret.md:198]: I'm fine to retain the existing wording (a later PR could update it), but if we want to change it, how about:

```suggestion
If you want to access data from a Secret in a Pod, one way to do that is to
have Kubernetes make the value of that Secret be available as a file inside
the filesystem of one or more of the Pod's containers.

To configure that, you:
```

File: content/en/docs/concepts/configuration/secret.md
Comment (sftim) [reply content/en/docs/concepts/configuration/secret.md:309]: Not convinced here. Is “beware” unidiomatic?

File: content/en/docs/concepts/configuration/secret.md
Comment (sftim) [reply content/en/docs/concepts/configuration/secret.md:539]: I'm very happy to save that for another PR - this one is making substantial enough changes that I don't see the need to cover every shortcoming we can spot.

File: content/en/docs/concepts/configuration/secret.md
Comment (sftim) [reply content/en/docs/concepts/configuration/secret.md:1036]:
```suggestion
However, using the defined and public Secret type (`kubernetes.io/basic-auth`) helps other
people to understand the purpose of your Secret, and sets a convention for what key names
to expect.
```

File: content/en/docs/concepts/configuration/secret.md
Comment (sftim) [inline content/en/docs/concepts/configuration/secret.md:1064]:
```suggestion
You could instead create an `Opaque` type Secret for credentials used for SSH authentication.
However, using the defined and public Secret type (`kubernetes.io/ssh-auth`) helps other
people to understand the purpose of your Secret, and sets a convention for what key names
to expect.
```

File: content/en/docs/concepts/configuration/secret.md
Comment (sftim) [reply content/en/docs/concepts/configuration/secret.md:1243]: Let's address this in a separate PR.

File: content/en/docs/concepts/configuration/secret.md
Comment (chrisnegus) [reply content/en/docs/concepts/configuration/secret.md:198]: I like that. It's more precise.

File: content/en/docs/concepts/configuration/secret.md
Comment (chrisnegus) [reply content/en/docs/concepts/configuration/secret.md:309]: I don't know that we would use the words differently on either side of the pond, but I leave it to you. I just thought that "beware" implies some danger while "be aware" is just something that would be helpful to know. If you think this facts represents danger, then beware is fine.

File: content/en/docs/concepts/configuration/secret.md
Comment (chrisnegus) [reply content/en/docs/concepts/configuration/secret.md:539]: Not a problem. You have already done a ton of work here.

File: content/en/docs/concepts/configuration/secret.md
Comment (chrisnegus) [reply content/en/docs/concepts/configuration/secret.md:1247]: Okay.

**Review by chrisnegus — COMMENTED** [review]
Really nice work @sftim and quick turn around. /lgtm as far as I'm concerned.

File: content/en/docs/concepts/configuration/secret.md
Comment (sftim) [reply content/en/docs/concepts/configuration/secret.md:309]: There is some danger here. People might specify the wrong `defaultMode` and leak a secret, because they thought that a number starting with `0` would be treated as octal.

**Comment by bradtopol** [comment]
Hi @sftim are you able to get the technical reviewers you need?

**Comment by sftim** [comment]
I don't think there's much here that needs technical review. Although it looks like there's new content, nearly all the text is based on the existing page but restructured.

**Comment by chrisnegus** [comment]
> I don't think there's much here that needs technical review. Although it looks like there's new content, nearly all the text is based on the existing page but restructured.

I agree that most of the PR reworks existing content and looks to me like it wouldn't break anything technical.

**Comment by chrisnegus** [comment]
@bradtopol Can you take a look at this and see if you think it can be merged as is?

**Comment by chrisnegus** [comment]
@sftim I see new commits from yesterday. Does this require further review or was it just housekeeping?

**Comment by sftim** [comment]
That was a no-op rebase.

**Comment by sftim** [comment]
When I've got open PRs that don't yet have LGTM, I try to rebase them every few months. The idea is to avoid merging in from a really stale merge base.

File: content/en/docs/concepts/configuration/secret.md
Comment (reylejano) [inline content/en/docs/concepts/configuration/secret.md:72]:
```suggestion
  [ServiceAccount](/docs/reference/access-authn-authz/authentication/#service-account-tokens)
```
nit: upper camel case?

**Comment by reylejano** [comment]
The preview link is updated:
https://deploy-preview-30984--kubernetes-io-main-staging.netlify.app/docs/concepts/configuration/secret/

Just a tiny suggestion which may not be correct otherwise LGTM
/lgtm

File: content/en/docs/concepts/configuration/secret.md
Comment (sftim) [reply content/en/docs/concepts/configuration/secret.md:72]: Sure.

**Comment by reylejano** [comment]
Re-applying lgtm
/lgtm

**Comment by chrisnegus** [comment]
@sftim It looks like it needs another rebase. You have done some really good work here. It would be sad to lose it.

**Comment by kbhawkey** [comment]
@sftim, Would you rebase? Thanks!

**Comment by sftim** [comment]
I do intend to rebase this; that's a chunky task so I'll have to find time to do so.

**Comment by sftim** [comment]
OK. Rebased!

**Comment by chrisnegus** [comment]
@kbhawkey @reylejano It looks like @sftim has done the rebase. Can this be merged?

**Review by onlydole — APPROVED** [review]
Thank you, @sftim 🥳

**Pull request merged**
